hoogle-test/src/Bug873/Bug873.hs

```diff
@@ -1,5 +0,0 @@
-module Bug873 (($), ($$)) where
```

Comment:
Is this test useless or something? Why is it removed? (I don’t object, just curious)

Reply:
Undecipherable failure, it's planned to reinstate it but we're short on contributors to properly take care of that.

Reply:
Looking into it

Reply:
By “undecipherable failure” you probably mean this:

```
Failed to run Haddock on test package 'Bug873'
hoogle-test: hoogle-test/out/Bug873: removeDirectoryRecursive:getSymbolicLinkStatus: does not exist (No such file or directory)
```

Turns out this error message is produced by the use of `removeDirectoryRecursive` in `haddock-test/src/Test/Haddock.hs`, in the `runHaddock` function:

```
        unless succeeded $ removeDirectoryRecursive (outDir cfgDirConfig tpkg)
```

This code is weird. It runs when the test failed (`unless succeeded`) and assumes that the `outDir` exists (`removeDirectoryRecursive` expects the directory to exist). However, it is precisely when the test fails that we don’t get an output directory. So we would get this undecipherable failure for any other failing test, not `Bug873` specifically. The test driver needs fixing.

By commenting out this line I got a more helpful error message:

```
Failed to run Haddock on test package 'Bug873'
Haddock output is at '/dev/null'. This file can be set with `--haddock-stdout`.
```

OK, so `haddock` fails on this file, but its output is hidden. I haven’t figured out how to use the `--haddock-stdout` option though, so I hacked it around by specifying `let haddockStdOut = stdout` in `runHaddock` and got this output:

```
haddock: While creating Haddock interface for Bug310:

While creating Haddock interface for Bug310:

get Wrap: Bad h
CallStack (from HasCallStack):
  error, called at src/Haddock/InterfaceFile.hs:729:12 in haddock-api-2.24.0-inplace:Haddock.InterfaceFile
```

Now we’re talking! This error is apparently produced by this code in `haddock-api/src/Haddock/InterfaceFile.hs`:

```
instance Binary DocName where
  put_ bh (Documented name modu) = do
    putByte bh 0
    put_ bh name
    put_ bh modu
  put_ bh (Undocumented name) = do
    putByte bh 1
    put_ bh name

  get bh = do
    h <- getByte bh
    case h of
      0 -> do
        name <- get bh
        modu <- get bh
        return (Documented name modu)
      1 -> do
        name <- get bh
        return (Undocumented name)
      _ -> error "get DocName: Bad h"
```

This is as far as I got. And since you mention binary serialization issues in https://github.com/haskell/haddock/pull/1296#issuecomment-766374759, I suspect it might be related.

Reply:
> OK, so haddock fails on this file, but its output is hidden. I haven’t figured out how to use the --haddock-stdout option though

I think you can pass it with `--test-options` from Cabal, or else if you `cabal run` the test suite, you can pass it to `--run-options`. The `removeDirectoryRecursive` is an issue I remember coming up against too (and I think I similarly would just change the code.

The binary serialization issue (for which I got a ping, but can't find the original comment), is probably due to the fact that the version of GHC used for CI comes with `.haddock` interface files built using the version of Haddock that was in the submodule at the time. However, this PR introduces breaking changes to that binary format. In order for this test to pass, you end up in a chicken and egg problem: you need GHC to bump the Haddock submodule so that the generated `base.haddock` matches the binary format that this PR has, but you can only bump the GHC submodule once this PR is merged.

The circular dependency problem is part of why at some point I embedded (most of) Haddock's test suite inside GHC's testsuite (at least there we would catch regressions). Over time, I've come more and more to believe that we should just try to move Haddock into GHC's tree (perhaps with the one exception being the `haddock-library` library, which is designed to build on older GHC versions, eg. so it can be used in Pandoc). Of course, the downside is that this would create lots of developer friction (in order to contribute to Haddock, you'd need to go through the whole GHC merge/build process).

Reply:
@harpocrates Hi! Thanks for chiming in, and sorry for the ghost ping, I had a bad case of "hands doing weird stuff with the GitHub interface".